File: scripts/kle2svg.ts
interface Key {
    x: number
    y: number
    w: number
}

interface State {
    row: number
    col: number
    x: number
    y: number
}

interface Modifier {
    x?: number
    y?: number
    w?: number
}

type Item = Modifier | string;
type KLE = Item[][];

function isModifier(mod: any): mod is Modifier {
    if (typeof mod === "string")
        return false;
    if ("x" in mod && typeof mod.x !== "number")
        return false;
    if ("y" in mod && typeof mod.y !== "number")
        return false;
    if ("w" in mod && typeof mod.w !== "number")
        return false;
    return true;
}

function isKle(data: any): data is KLE {
    if (data === null || data === undefined || !(data instanceof Array)) {
        return false;
    } else {
        for (const x of data) {
            if (!(x instanceof Array))
                return false;
            let wasMod = false;
            for (const y of x) {
                if (typeof y === "string") {
                    wasMod = false;
                } else if (!wasMod && isModifier(y)) {
                    wasMod = true;
                } else {
                    return false;
                }
            }
            if (wasMod)
                return false;
        }
    }
    return true;
}

async function readJson(path: string): Promise<null | KLE> {
    const text = await Deno.readTextFile(path);
    const data = JSON.parse(text);
    return isKle(data) ? data : null;
}

function nextKey(s: State, data: KLE): null | [State, Key] {
    const isLast = data[s.row].length == s.col;
    const { row, col, x, y } = isLast
        ? { row: s.row + 1, col: 0, x: 0, y: s.y + 1 }
        : s;

    if (data.length === row)
        return null;

    const v = data[row][col];
    const [newCol, m] = typeof v === "string"
        ? [col + 1, { x: 0, y: 0, w: 1 }]
        : [col + 2, { x: 0, y: 0, w: 1, ...v }];
    return [
        { row, col: newCol, x: x + m.x + m.w, y: y + m.y },
        { x: x + m.x, y: y + m.y, w: m.w }
    ];
}

function* iterKeys(data: KLE) {
    const s0: State = { row: 0, col: 0, x: 0, y: 0 };
    let res = nextKey(s0, data);
    while (res) {
        const [newState, key] = res;
        yield key;
        res = nextKey(newState, data);
    }
}

const s0: State = {
    row: 0,
    col: 0,
    x: 0,
    y: 0,
};

interface Show {
    toString(): string
}

interface Node {
    name: string
    attrs: Record<string, Show>
    children: Node[]
}

function makeAttrs(attrs: Record<string, Show>) {
    return Object.entries(attrs)
        .map(([k, v]) => `${k}="${v}"`).join(" ");
}

function nodes2xml(nodes: Node[]): string {
    return nodes.map(n => `<${n.name} ${makeAttrs(n.attrs)}>${nodes2xml(n.children)}</${n.name}>`).join("\n");
}

function keyToSvg({x, y, w}: Key): Node[] {
    const U = 19.05;
    return [
        {
            name: "rect",
            attrs: {
                "x": x * U,
                "y": y * U,
                "width": w * U,
                "height": U,
                "stroke": "black",
                "stroke-width": 0.2,
                "fill-opacity": 0.1,
            },
            children: []
        },
        {
            name: "rect",
            attrs: {
                "x": (x + w/2) * U - 7,
                "y": y * U + (U - 14) / 2,
                "width":  14,
                "height": 14,
                "fill-opacity": 0.3,
            },
            children: []
        },
        {
            name: "circle",
            attrs: { cx: (x + w/2) * U - 3.81, cy: (y + 0.5) * U - 2.54, r: 1.5 },
            children: []
        },
        {
            name: "circle",
            attrs: { cx: (x + w/2) * U + 2.54, cy: (y + 0.5) * U - 5.08, r: 1.5 },
            children: []
        },
        {
            name: "circle",
            attrs: { cx: (x + w/2) * U, cy: (y + 0.5) * U, r: 2 },
            children: []
        },
    ];
}

const path = "undiscipline.json";
const data = await readJson(path);
if (data) {
    const svg: Node[] = Array.from(iterKeys(data)).flatMap(keyToSvg);
    Deno.writeTextFile("undiscipline.svg", (
        '<?xml version="1.0" encoding="UTF-8" ?>'
        + '<svg width="800" viewBox="0 0 350 150" xmlns="http://www.w3.org/2000/svg">'
        + nodes2xml(svg)
        + '</svg>\n'
    )).then(() => console.log("Done!"));
} else {
    console.error(`Could not open file ${path}`);
}

File: scripts/kle2svg.py
import json
from itertools import chain
from dataclasses import dataclass, field
from decimal import Decimal
from typing import Any, Optional, Union, Iterable


@dataclass
class Key:
    x: Decimal
    y: Decimal
    w: Decimal
    ith: int


@dataclass
class State:
    ith: int
    row: int
    col: int
    x: Decimal
    y: Decimal


@dataclass
class Modifier:
    x: Decimal
    y: Decimal
    w: Decimal


Item = Union[Modifier, str]
KLE = list[list[Item]]


def parse_item(mod: Any) -> Optional[Item]:
    if type(mod) == str:
        return mod
    if type(mod) == dict:
        return Modifier(
            mod.get("x", Decimal(0)), mod.get("y", Decimal(0)), mod.get("w", Decimal(1))
        )
    return None


def parse_kle(data: Any) -> Optional[KLE]:
    kle: KLE = []
    if type(data) != list:
        return None
    for inner in data:
        if type(inner) != list:
            return None
        kle.append([])
        for x in inner:
            item = parse_item(x)
            if item is None:
                return None
            kle[-1].append(item)
    return kle


def parse_json(path: str) -> Optional[KLE]:
    with open(path) as f:
        data = json.load(f, parse_float=Decimal, parse_int=Decimal)
    return parse_kle(data)


def next_key(s: State, data: KLE) -> Optional[tuple[State, Key]]:
    if len(data[s.row]) == s.col:
        row, col, x, y = s.row + 1, 0, 0, s.y + 1
    else:
        row, col, x, y = s.row, s.col, s.x, s.y

    if len(data) == row:
        return None

    v = data[row][col]
    if type(v) == str:
        newCol = col + 1
        m = Modifier(x=Decimal(0), y=Decimal(0), w=Decimal(1))
    else:
        newCol = col + 2
        m = v
    return (State(s.ith + 1, row, newCol, x + m.x + m.w, y + m.y), Key(x + m.x, y + m.y, m.w, s.ith))


def iter_keys(data: KLE):
    s0 = State(1, 0, 0, Decimal(0), Decimal(0))
    res = next_key(s0, data)
    while res:
        newState, key = res
        yield key
        res = next_key(newState, data)

Content = Union[None, list["Node"], str]

@dataclass
class Node:
    name: str
    attrs: dict[str, Any] = field(default_factory=dict)
    content: Content = None


def make_attrs(attrs: dict[str, Any]) -> str:
    return " ".join(map(lambda x: f'{x[0]}="{x[1]}"', attrs.items()))


def nodes2xml(nodes: Iterable[Node]) -> str:
    def render_content(content: Content) -> str:
        if content is None:
            return ""
        elif isinstance(content, str):
            return content
        else:
            return nodes2xml(content)
    txts = (
        f"<{n.name} {make_attrs(n.attrs)}>{render_content(n.content)}</{n.name}>"
        for n in nodes
    )
    return "\n".join(txts)


def key2svg(key: Key) -> list[Node]:
    x = key.x
    y = key.y
    w = key.w
    U = Decimal("19.05")
    print(f"key {key.ith}:", (x + w / Decimal("2")) * U * 1000000, (y + Decimal("0.5")) * U * 1000000)
    return [
        Node(
            name="rect",
            attrs={
                "x": x * U,
                "y": y * U,
                "width": w * U,
                "height": U,
                "stroke": "black",
                "stroke-width": 0.2,
                "fill": "#eee",
            },
        ),
        Node(
            name="rect",
            attrs={
                "x": (x + w / 2) * U - 7,
                "y": y * U + (U - 14) / 2,
                "width": 14,
                "height": 14,
                "fill": "#aaa",
            },
        ),
        Node(
            name="circle",
            attrs={
                "cx": (x + w / 2) * U - Decimal("3.81"),
                "cy": (y + Decimal("0.5")) * U - Decimal("2.54"),
                "r": 1.5,
            },
        ),
        Node(
            name="circle",
            attrs={
                "cx": (x + w / 2) * U + Decimal("2.54"),
                "cy": (y + Decimal("0.5")) * U - Decimal("5.08"),
                "r": 1.5,
            },
        ),
        Node(
            name="circle",
            attrs={"cx": (x + w / 2) * U, "cy": (y + Decimal("0.5")) * U, "r": 2},
        ),
        Node(
            name="text",
            attrs={
                "text-anchor": "end",
                "font-size": "4pt",
                "fill": "red",
                "x": (x+ w / 2) * U + 7,
                "y": y * U + 16
            },
            content=str(key.ith),
        )
    ]


if __name__ == "__main__":
    path = "undiscipline.json"
    data = parse_json(path)
    if data:
        svg = chain.from_iterable(map(key2svg, iter_keys(data)))
        with open("undiscipline.svg", "w") as f:
            f.write('<?xml version="1.0" encoding="UTF-8" ?>')
            f.write(
                '<svg width="420mm" height="297mm" viewBox="0 0 420 297" xmlns="http://www.w3.org/2000/svg">'
            )
            f.write(nodes2xml(svg))
            f.write("</svg>")
    else:
        print(f"Could not open file {path}")
